docs/historical/mh-jun-1982/man/rmail.1

.TH RMAIL 1 RAND
.SH NAME
rmail \- send mail from a remote site
.SH SYNOPSIS
.B mail
user 
.SH DESCRIPTION
.I Rmail
is intended as an interface between mail sent from foreign
sites via
.I uucp(1)
and the Rand message handling system,
.I mh(1).
It reads a message from the standard input and, after creating
mh-format lines for the sender and recipient user, writes
it into a temporary file.  It then invokes
.I deliver(8)
to complete the mail delivery.
.PP
.I Rmail
is able to understand a series of lines of the form:
.br
	From \fIuser\fB ... remote from \fIsite\fR
.br
and turn them into the correct "From: \fIsite\fR!\fIuser\fR" form,
thus allowing effective use of the 
.I reply(1)
command.
.PP
If the destination 
.I user
is in fact a foreign uucp site, no formatting or reprocessing is
done; 
.I deliver
will simply forward the message to the foreign site with the
appropriate uucp-style header on it.
.SH FILES
/etc/mh/deliver
.br
/tmp/mail*.tmp
.br
.SH SEE ALSO
bellmail(1),  deliver(8), mail(1), uucp(1)


NEWS
# NEWS - Release notes for nmh 1.7
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been over three years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time
MH and nmh uses should read careful the NOTEABLE CHANGES section, as there
are some significant changes to nmh behavior.  Otherwise, please see the
README and INSTALL files for help on getting started with nmh.

For news of future releases, subscribe to the low-volume
    https://lists.nongnu.org/mailman/listinfo/nmh-announce

---------------
NOTABLE CHANGES
---------------

The largest notable changes in the 1.7 release are:

- Complete unification of network security support.  All network protocols
  (currently, POP and SMTP) have been refactored to use a common set of
  security routines.  This means all protocols support all SASL mechanisms
  (via the Cyrus-SASL library) and TLS.  TLS support has been strengthened
  to perform certificate name validation and to require TLS 1.1 as a
  minimum protocol.  Also, all protocols can make use of the OAuth2/XOAUTH
  SASL mechanism, which is supported by Gmail.
- send(1) now supports adding switches to post(8) based on the address or
  domain of the email address in the From: header; this more easily allows
  users to support multiple identities.
- A generic facility for passing arguments to filter programs in repl(1)
  by use of the -convertargs switch.
- Native support for the manipulation of iCalendar requests; see mhical(1)
  for more details.

------------
NEW FEATURES
------------

The following are new features for the 1.7 release of nmh:

- When building from source, configure will derive ${prefix} from an existing
  nmh installation if it finds one in your $PATH.
- Added welcome message when nmh detects that its version changed.
- The default locations for configuration files and support binaries
  have been changed.  Configuration files now install into ${sysconfdir}/nmh,
  and support binaries are placed in ${libexecdir}/nmh.  If you are upgrading
  an existing installation you should look for old configuration files in
  ${sysconfdir} and merge any local customizations into the new files in
  ${sysconfdir}/nmh, then remove the old files.  ${libdir} will also contain
  obsolete support programs that should be removed.
- All TLS connections now perform certificate validation (including hostname
  matching) by default; can be disabled on a per-application basis.
- post now defaults to port 587 on 'smtp' message submission.
- A value of 0 for the width switch of scan(1), inc(1), ap(1), dp(1),
  fmttest(1), and mhl(1) now means as many characters as the format
  engine can produce [Bug #15274].  That amount is limited by internal
  buffers.
- If a component has trailing whitespace, e.g., body:component="> ",
  mhl now trims that whitespace off when filtering blank text lines.
- An "rtrim" flag has been added to mhl to remove any trailing
  whitespace from filtered text lines.  A corresponding "nortrim" flag
  has also been added.
- Added getmymbox and getmyaddr mh-format(5) function escapes.
- New -[no]changecur, -fixtype, -decodetypes, and -[no]crlflinebreaks switches
  have been added to mhfixmsg(1).
- mhfixmsg now removes an extraneous trailing semicolon from header
  parameter lists.
- Added -convertargs switch to repl(1), to pass arguments to programs
  specified in the user's profile or mhn.defaults to convert message
  content.
- Added mhical(1), to display, reply to, and cancel iCalendar (RFC 5545)
  event requests.
- Added multiply mh-format(5) function.
- "mhparam bindir" prints the path to the directory containing the public
  executables (${bindir}).
- New "-prefer" switch for mhshow (and mhlist and mhshow), to allow specifying
  the preferred content types to show, if present in a multipart alternative.
- mh-format now has %(kilo) and %(kibi) functions, to allow printing
  numbers with SI or IEC quantities, e.g. "10K", "2.3Mi".
- Support for the -sendmail flag to send/post to change the sendmail
  binary when using the sendmail/pipe MTS.
- Added support to send(1) to specify switches to post(1) based on address or
  domain name in From: header line in message draft.
- post(8) -snoop now attempts to decode base64-encoded SMTP traffic.
- folder(1) -nocreate now prints a warning message for a non-existent folder.
- mhfixmsg(1) now allows -decodetext binary, though 8bit is still the default.
- inc(1) and msgchk(1) now support TLS encryption natively.
- All network protocols support the XOAUTH authentication mechanism.
- Support for SMTPUTF8 (RFC 6531) has been added.  mhshow(1) already supported
  RFC 6532, assuming all 8-bit message header field bodies are UTF-8 and use
  of a UTF-8 locale.
- mhfixmsg now replaces RFC 2047 encoding with RFC 2231 encoding of name and
  filename parameters in Content-Type and Content-Disposition headers,
  respectively.
- If a message body contains 8-bit bytes, post(8) uses SMTP 8BITMIME if the
  server supports it.  If not, post fails with a message to the user to
  encode the message for 7-bit transport.
- Fewer lseek(2)s will be used when reading headers in the common case.
- ./configure's --enable-debug has been removed;  it did nothing.
- configure now defaults to enabling each of TLS and Cyrus SASL if the
  necessary headers and libraries are found.
- Moved build_nmh to top-level directory.
- Better error reporting for connections to network services.

---------
BUG FIXES
---------

- The format scanner no longer subtracts 1 from the width.  This has the
  effect of no longer counting the trailing newline in the output of
  scan(1), inc(1), and the other programs that rely on the format scanner.
- The first character of some very short (less than 4 characters) message
  bodies is no longer dropped.
- Single-character headers can be reliably formatted, etc., instead of
  apparently being missing.

docs/historical/mh-jun-1982/progs/rmail.c
#ifdef COMMENT
	Proprietary Rand Corporation, 1981.
	Further distribution of this software
	subject to the terms of the Rand
	license agreement.
#endif

static char *sccsid = "@(#)from rmail.c      4.1 (Berkeley) 10/1/80";
/*
 * rmail: front end for mail to stack up those stupid >From ... remote from ...
 * lines and make a correct return address.  This works with the -f option
 * to /etc/delivermail so it won't work on systems without delivermail.
 * However, it ought to be easy to modify a standard /bin/mail to do the
 * same thing.
 *
 * NOTE: Rmail is SPECIFICALLY INTENDED for ERNIE COVAX because of its
 * physical position as a gateway between the uucp net and the arpanet.
 * By default, other sites will probably want /bin/rmail to be a link
 * to /bin/mail, as it was intended by BTL.  However, other than the
 * (somewhat annoying) loss of information about when the mail was
 * originally sent, rmail should work OK on other systems running uucp.
 * If you don't run uucp you don't even need any rmail.
 *
 * This version revised around New Year's Day 1981 to interface with
 * Rand's MH system.  The nature of the revision is to (1) do nothing
 * if the destination is uucp-remote (note the assumption of single
 * destination) and (2) to call "/etc/mh/deliver" to do the deed.
 *
 * Revised Aug 1981 to add NOGATEWAY screen.  Also fixed a few bugs.
 * PK.
 *
 * 5/8/82: Tack on uu-Date: if msg has no date.  PK.
 */

/*#define DEBUG 1*/

#include "../mh.h"
#include <whoami.h>
#include <stdio.h>

#define MATCH   0
#define PARSE   1
#define NOPARSE 0


char *index();

FILE *out;              /* output to delivermail */
char *tmpfil;           /* file name of same */
char tmpfila[24];       /* array for tmpfil */
char *to;               /* argv[1] */
char from[512];         /* accumulated path of sender */
char lbuf[512];         /* one line of the message */

char d1[10], d2[10], d3[10], d4[10], d5[10];   /*** ctime() fields ***/

#ifdef ARPANET
int netaddr;            /* set if "to" contains an arpanet address */
#endif

main(argc, argv)
char **argv;
{
	char ufrom[64];	/* user on remote system */
	char sys[64];	/* a system in path */
	char junk[512];	/* scratchpad */
	char *cp;
	int badhdr;   /***/

	to = argv[1];
	if (argc != 2) {
		fprintf(stderr, "Usage: rmail user\n");
		exit(1);
	}
#ifdef DEBUG
	out=stdout;
	tmpfil = "/dev/tty";
#else
	tmpfil = tmpfila;
	sprintf (tmpfil, "/tmp/%s", makename ("mail",".tmp"));
	if ((out=fopen(tmpfil, "w")) == NULL) {
		fprintf(stderr, "Can't create %s\n", tmpfil);
		exit(1);
	}
#endif
	for (;;) {
		fgets(lbuf, sizeof lbuf, stdin);
		if (strncmp(lbuf, "From ", 5) && strncmp(lbuf, ">From ", 6))
			break;
		fputs(lbuf, out); /* Save--in case we are just forwarding */
/***/           sscanf(lbuf, "%s %s %s %s %s %s %s remote from %s",
		 junk, ufrom, d1, d2, d3, d4, d5, sys);
/*              sscanf(lbuf, "%s %s", junk, ufrom);  */
		cp = lbuf;
		for (;;) {
			cp = index(cp+1, 'r');
			if (cp == NULL)
				cp = "remote from somewhere";
#ifdef DEBUG
			printf("cp='%s'\n", cp);
#endif
			if (strncmp(cp, "remote from ", 12) == MATCH)
				break;
		}
		sscanf(cp, "remote from %s", sys);
		strcat(from, sys);
		strcat(from, "!");
#ifdef DEBUG
		printf("ufrom='%s', sys='%s', from now '%s'\n", ufrom, sys, from);
#endif
	}
	strcat(from, ufrom);

#ifdef DEBUG
	printf("from now '%s'\n",  from);
#endif


#ifdef ARPANET
	netaddr = isarpa(to);           /* Arpanet destination?  */

#ifdef NOGATEWAY
	if(netaddr && !okhost(sys))
	{
		truncate();
		returnmail("Sorry, not an Arpanet gateway!");
		exit(0);
	}
#endif

	if (index (to, '!') && !netaddr)  {
#else
	if (index (to, '!')) {
#endif
		/* Just forwarding! */
		putmsg(NOPARSE);
		deliver(to);
		exit(0);
	}

	truncate();
	/* fprintf(out, "To: %s\n",to); */
	putfrom();

	if( !((cp = index(lbuf, ':')) &&  (cp - lbuf <  NAMESZ ))) {
		fputs("\n",out);/* Doesn't look good; terminate hdr */
		badhdr++;
	}
	putmsg(badhdr?NOPARSE:PARSE);
	deliver(to);

}


deliver(to)
	char *to;
{
	int sts,pid,waitid;

#ifdef DEBUG
	printf("%s would get called here; Delivery to: %s \n", mh_deliver,to);
	exit(0);
#endif
	fclose(out);

	if ((pid=fork()) == -1) {
		fprintf(stderr, "Cannot fork Deliver!\n");

#ifndef DEBUG   /* Extra precaution */
		unlink(tmpfil);
#endif
		exit(1);
	}
	if(pid) {
		while(((waitid = wait(&sts)) != pid) && (waitid != -1));

#ifndef DEBUG   /* Extra precaution */
		unlink(tmpfil);
#endif
		exit(0);
	}
	execl(mh_deliver, "deliver", "-deliver", to, tmpfil, 0);

	perror( "Cannot exec Deliver ");

	exit(1);

}

#ifdef ARPANET

isarpa(str)                             /* Gateway to Arpanet? */
	char *str;
{
	char *cp;

	if (index (str, '@'))
		return(1);

	for (cp = str; ;) {
		if((cp = index(cp, ' ')) == NULL)
			return(0);
		while(*cp == ' ')
			cp++;
		if ((strncmp(cp, "at ", 3) == MATCH) ||
		    (strncmp(cp, "AT ", 3) == MATCH) ||
		    (strncmp(cp, "At ", 3) == MATCH))

			return(1);
	}
}


okhost(str)             /* Host permitted to use us as an arpanet gateway? */
	char *str;
{
	register short i;

	for (i=0; rhosts[i]; i++)
		if (strcmp (str, rhosts[i]) == MATCH)
			return(1);
	return(0);
}
#endif

returnmail(message)
	char  *message;
{

#ifdef DEBUG
	printf("returnmail()\n");
#endif
	if(!from)  return;

	putdate();
	fprintf(out, "To: %s\n", from);
	fputs("\n",out);
	fputs (message, out);
	fputs("\n\n\n*--------------RETURNED MESSAGE---------------*\n\n",
	  out);

	putmsg(NOPARSE);
	deliver(from);

}


#include <sys/types.h>
#include <sys/timeb.h>
#include <time.h>

putdate()
{
	long  now;
	register char *t, *p;
	char *timezone();
	struct timeb tb;
	struct tm *tmp;
	static char *wday[] = {
		"Sun",
		"Mon",
		"Tues",
		"Wednes",
		"Thurs",
		"Fri",
		"Satur"
	};

	now = time((long *)0);
	t = ctime(&now);
	ftime(&tb);
	tmp = localtime(&now);
	p = timezone(tb.timezone, tmp->tm_isdst);

	/*                  day    13   Apr  1981 20  :38   -PST */
	fprintf(out, "Date: %sday, %.2s %.3s %.4s %.2s:%.2s-%.3s\n",
		     wday[tmp->tm_wday], t+8, t+4, t+20, t+11, t+14, p);
}


putfrom()
{
	if (strlen(from)) 
#ifdef ARPANET
		if (netaddr)
			fprintf(out, "From: %s at %s\n",from, HOSTNAME);
		else
			fprintf(out, "From: %s\n", from);
#else
		fprintf(out, "From: %s\n",from);
#endif

}


putmsg(parse)
int parse;
{
	int dateseen = 0;

	if(!parse)
		putall();
	else {
		if (uleqn(lbuf, "date:", 5) == 0)
			dateseen++;
		fputs(lbuf, out);
		while (fgets(lbuf, sizeof lbuf, stdin)) {
			if(lbuf[0] == '\n' ) { /* end of hdrs */
				if(!dateseen)
					uudate();
				putall();
			} else {
				if (uleqn(lbuf, "date:", 5) == 0)
					dateseen++;
				fputs(lbuf, out);
			}
		}
	}
}


putall()
{
	fputs(lbuf, out);
	while (fgets(lbuf, sizeof lbuf, stdin))
		fputs(lbuf, out);
}


truncate()
{
	/* Truncate those "...remote from..." header lines. */
	/* They're kept only if the message is to be uucp-forwarded */

#ifndef DEBUG
	fclose (out);   out=fopen(tmpfil, "w");
#endif

}


/*
 * Compare strings (at most n bytes) without regard to case.
 *   Returns:   s1>s2: >0,  s1==s2: 0,  s1<s2: <0.
 */

uleqn(s1, s2, n)
register char *s1, *s2;
register n;
{

	while (--n >= 0 && (*s1|040) == (*s2|040)) {
		s2++;
		if (*s1++ == '\0')
			return(0);
	}
	return(n<0 ? 0 : (*s1|040) - (*s2|040));
}


uudate()
{
	char *prefix();

	/*                  day    13   Apr  1981 20  :38   -PST */
	fprintf(out, "Date: %sday, %.2s %.3s %.4s %.2s:%.2s-%.3s\n",
		     prefix(d1), d3, d2, d5, d4, d4+3, "???");
}

char *
prefix(str)
char *str;
{
	static char *wday[] = {
		"Sun",
		"Mon",
		"Tues",
		"Wednes",
		"Thurs",
		"Fri",
		"Satur",
		0
	};

	register char **wp;

	for(wp=wday; wp; wp++)
		if(uleqn(str, *wp, 3) == 0)
			return(*wp);
	return("???");
}
